tfjs-backend-webgpu/src/shader_preprocessor_wgsl.ts

```diff
@@ -0,0 +1,604 @@
+        return get${texFuncSnippet}();
+      }
+    `;
+  } else {
+    if (outRank < 2 && broadcastDims.length >= 1) {
+      coordsSnippet = 'coords = 0u;';
+    } else {
+      coordsSnippet =
+          broadcastDims.map(d => `coords[${d + rankDiff}u] = 0u;`).join('\n');
+    }
+  }
+
+  let unpackedCoordsSnippet = '';
+  if (outRank < 2 && inRank > 0) {
+    unpackedCoordsSnippet = 'coords';
+  } else {
+    if (outRank > 1) {
+      const coordsType = getCoordsDataTypeWgsl(inRank);
+      const coordsValues =
+          inInfo.shape.map((s, i) => `coords[${i + rankDiff}u]`).join(', ');
+      unpackedCoordsSnippet = `${coordsType}(${coordsValues})`;
+    } else {
+      unpackedCoordsSnippet = 'coords';
+    }
+  }
+
+  const shapeStr =
+      `uniforms.${texName.charAt(0).toLowerCase() + texName.slice(1)}Shape`;
+  const rankStr = `${inRank}D`;
+  if (isVec4) {
+    return `
+      fn ${funcName}ByGlobalId(globalId : vec3<u32>) -> vec4<f32> {
+        var coords : ${type} = getOutputCoords(globalId);
+        ${coordsSnippet}
+        return ${texName}.numbers[getFlatIndex${rankStr}(${
+        unpackedCoordsSnippet}, ${shapeStr}) / 4u];
+      }
+
+      fn ${funcName}ByCoords(coordsIn : ${type}) -> vec4<f32> {
+        var coords : ${type} = coordsIn;
+        ${coordsSnippet}
+        return ${texName}.numbers[getFlatIndex${rankStr}(${
+        unpackedCoordsSnippet}, ${shapeStr}) / 4u];
+      }
+    `;
+  }
+
+  return `
+    fn ${funcName}ByGlobalId(globalId : vec3<u32>) ->f32 {
+      var coords : ${type} = getOutputCoords(globalId);
+      ${coordsSnippet}
+      return f32(${texName}.numbers[getFlatIndex${rankStr}(${
+      unpackedCoordsSnippet}, ${shapeStr})]);
+    }
+
+    fn ${funcName}ByCoords(coordsIn : ${type}) ->f32 {
+      var coords : ${type} = coordsIn;
+      ${coordsSnippet}
+      return f32(${texName}.numbers[getFlatIndex${rankStr}(${
+      unpackedCoordsSnippet}, ${shapeStr})]);
+    }
+  `;
+}
+
+/**
+ * Generates getOutputCoords() function that computes output coordinates from
+ * dispatch geometry to reduce arithmetic.
+ */
+export function generateGetOutputCoords(
+    outShape: number[],
+    dispatchLayout: {x: number[], y?: number[], z?: number[]}):
+    [string, number] {
+  const {x, y = [], z = []} = dispatchLayout;
+
+  const outRank = outShape.length;
+  if (x.length === outRank) {
+    const dtype = getCoordsDataTypeWgsl(outRank);
+    const snippet = `fn getOutputCoords(globalId : vec3<u32>) -> ${dtype}{
+      return getCoordsFromFlatIndex(u32(globalId.x));
+    }
+    `;
+    return [snippet, outRank];
+  }
+
+  let gatherDimensionsStr = '';
+  const dims = [x, y, z];
+
+  let rank = 0;
+
+  for (let i = 0; i < dims.length; i++) {
+    const arr = dims[i];
+
+    if (arr.length === 0) {
+      continue;
+    }
+
+    rank += arr.length;
+
+    if (arr.length === 1) {
+      gatherDimensionsStr += `let d${arr[0]} : u32 =
+        u32(globalId[${i}]);`;
```

Comment:
`= globalId[${i}];`  `u32(...)` is not needed?

Reply:
done